src/cipher/random.ml

```diff
@@ -8,40 +8,22 @@
 (* TODO LATER Is there something with lighter dependencies? Although perhaps
    these are not so bad... *)
 
-exception Entropy_is_not_initialized
-
-let setup_entropy =
-  "\nTo initialize entropy with a default random number generator, and \
-   set up Dream, do the following:\
-   \n  If you are using Lwt/Unix, execute `Dream.random_initialize ()`
-   \n  If you are using MirageOS, use the Dream device in config.ml
-   \n"
-
-let () = Printexc.register_printer @@ function
-  | Entropy_is_not_initialized ->
-    Some ("The entropy is not yet initialized. " ^ setup_entropy)
-  | _ -> None
-
 let set = ref false
 
 let _initialized = ref None
 
 let initialized () : [ `Initialized ] =
   match !_initialized with
-  | None -> raise Entropy_is_not_initialized
+  | None -> failwith "Entropy is not initialized."
   | Some v -> Lazy.force v
 
 let initialize f =
-  if !set then Logs.debug (fun log -> log
-    "Dream__cipher.Random.initialize has already been called, ignoring this call.")
-  else begin
+  if not !set
+  then begin
     ( try
-        let `Initialized = initialized () in
-        Format.eprintf
-          "Dream__cipher.Random.initialize has already been set, check that this call \
-          is intentional";
+        let `Initialized = initialized () in ()
         with
-          Entropy_is_not_initialized -> ());
+          Failure _ -> ());
     set := true ;
     _initialized := Some (Lazy.from_fun f)
```

Comment:
this function has rather complex semantics, it should just "_initialized := Some (Lazy.from_fun f)".

as it is now, it first calls to `initialized` (why?) - which always fails! and then sets two reference cells.

it reminds me a lot on the initialization in Mirage_crypto_rng -- which is mainly there to avoid pure unix and lwt mixing (+ informative messages if these are mixed in a binary).

Reply:
(the semantics is tricky since the case where `_initialized` is already `Some _`, you'll force the value here.